src/main/java/org/opensearch/knn/index/codec/transfer/OffHeapVectorTransfer.java

```diff
@@ -27,16 +28,22 @@ public abstract class OffHeapVectorTransfer<T> implements Closeable {
 
     @Getter
     private long vectorAddress;
+    @Getter
     protected final int transferLimit;
 
-    private final List<T> vectorsToTransfer;
+    private List<T> vectorsToTransfer;
 
-    public OffHeapVectorTransfer(final int transferLimit) {
-        this.transferLimit = transferLimit;
-        this.vectorsToTransfer = new ArrayList<>(transferLimit);
+    public OffHeapVectorTransfer(int bytesPerVector, int totalVectorsToTransfer) {
+        this.transferLimit = computeTransferLimit(bytesPerVector, totalVectorsToTransfer);
+        this.vectorsToTransfer = new ArrayList<>(this.transferLimit);
         this.vectorAddress = 0;
     }
 
+    private int computeTransferLimit(int bytesPerVector, int totalVectorsToTransfer) {
+        int limit = (int) Math.max(1, KNNSettings.getVectorStreamingMemoryLimit().getBytes() / bytesPerVector);
+        return Math.min(limit, totalVectorsToTransfer);
+    }
```

Comment:
[Not a blocker for this PR to be merged]
lets have some java doc on this explaining this logic.